web_api.md

```diff
@@ -0,0 +1,265 @@
+# Web API
+
+## Why do we need a Web API?
+
+After having processors running locally via the [CLI](https://ocr-d.de/en/spec/cli), communication over network is the
+natural extension. This feature will improve the flexibility, scalability and reliability of the system. This
+specification presents ideas behind the endpoints, how to use them, and technical details happening in the background.
+
+## The Specification
+
+The Web API specification can be found [here](openapi.yml). It follows
+the [OpenAPI specification](https://swagger.io/specification/). There are 4 parts to be implemented: discovery,
+processing, workflow, and workspace.
+
+**Discovery**: The service endpoints in this section provide information about the server. They include, but are not
+limited to, hardware configuration, installed processors, and information about each processor.
+
+**Processing**: Via the service endpoints in this section, one can get information about a
+specific [processor](https://ocr-d.de/en/spec/glossary#ocr-d-processor), trigger a processor run, and check the status
+of a running processor. By exposing these endpoints, the server can encapsulate the detailed setup of the system and
+offer users a single entry to the processors. The implementation of this section is provided
+by [OCR-D/core](https://github.com/OCR-D/core). Implementors do not need to implement it themselves, they can reuse
+and/or extend the reference implementation from OCR-D/core.
+
+**Workflow**: Beyond single processors, one can manage
+entire [workflows](https://ocr-d.de/en/spec/glossary#ocr-d-workflow), i.e. a series of connected processor
+instances. In this spec, a workflow amounts to a [Nextflow](https://www.nextflow.io/) script. Some information
+about Nextflow and how to use it in OCR-D is documented [in the Nextflow spec](nextflow).
+
+**Workspace**: The service endpoints in this section concern data management, which in OCR-D is handled
+via [workspaces](https://ocr-d.de/en/spec/glossary#workspace). Processing (via single processors or workflows) always
+refers to existing workspaces, i.e. combinations of `mets.xml` and downloaded files from the `mets.xml`.
```

Comment:
I didn't add this case since I want to leave this endpoint optional. Some IPs may want to support METS URL, some don't. And it also depends on where the METS files are stored, some kind of credentials might be necessary. That's why I leave it to the IP to decide.

Reply:
> I didn't add this case since I want to leave this endpoint optional. Some IPs may want to support METS URL, some don't. 

How about an **optional** endpoint then? IMO this should not be left underspecified (because it helps understanding the whole thing and because a standard endpoint would be preferable).

> And it also depends on where the METS files are stored, some kind of credentials might be necessary.

What do you mean by that?

Reply:
Since I thought that all endpoints defined in the spec are compulsory. But maybe I can add some text to note that this endpoint is optional.

METS files are not always publicly accessible, right? There might be cases where authentication is needed to download the files.

Reply:
> Since I thought that all endpoints defined in the spec are compulsory.

The CLI spec has optional features, too.

> METS files are not always publicly accessible, right? There might be cases where authentication is needed to download the files.

Ah, ok. But why should that hinder establishing the endpoint itself? Even if an implementation needs to add credentials, or the URLs are not valid to some outside network, the server can still build on the same implementation.

Reply:
Since I didn't know how to detail describe the `credentials` part to cover all cases, I think it's best to leave it out. But never mind, as you said, we can have this endpoint defined and mark it as optional.